### PR TITLE
Add support for solution pool of MIP solvers in MindtPy

### PR DESCRIPTION
**Pull request by ZedongPeng**
## Summary/Motivation:
This PR adds the support for solution pool of CPLEX and GUROBI in MindtPy. With the help of solution pool, MindtPy now can explore several integer combinations in one iteration. 

## Changes proposed in this PR:
Two new config options are created.

```
CONFIG.declare('solution_pool', ConfigValue(
    default=False,
    description='Use solution pool in solving the MILP main problem.',
    domain=bool
))
CONFIG.declare('num_solution_iteration', ConfigValue(
    default=5,
    description='The number of MIP solutions (from the solution pool) used to generate the fixed NLP subproblem in each iteration.',
    domain=PositiveInt
))
```

## Need to discuss
The value solution-pool-related parameters in CPLEX and GUROBI, like intensity and capacity.
Now, all these parameters are set as default.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by ZedongPeng** [comment]
Hi John @jsiirola. Function `copy_var_list_values_from_solution_pool` will extract the variable values from the solution pool interface. However, I didn't find a way to set or change these values. For example, when using `gurobi_persistent`, the following error will show up.
> `AttributeError: Attribute 'Xn' cannot be set.`

They seem to be read-only. Therefore, I removed the redundant code in `copy_var_list_values_from_solution_pool` function, which will also improve the code coverage.

File: pyomo/contrib/mindtpy/util.py
Comment (mrmundt) [inline pyomo/contrib/mindtpy/util.py:625]: Typo: `necessary`

File: pyomo/contrib/mindtpy/util.py
Comment (mrmundt) [inline pyomo/contrib/mindtpy/util.py:626]: Typo: `necessary`

**Review by mrmundt — CHANGES_REQUESTED** [review]

File: pyomo/contrib/mindtpy/util.py
Comment (jsiirola) [inline pyomo/contrib/mindtpy/util.py:665]: Generally, this should raise an instance of a ValueError and not the ValueError class.  Further, as you are already in an exception that is raising a ValueError, I would recommend logging a message and re-raising the original exception:
```python
else:
    config.logger.error('Unknown validation domain error setting variable %s', (v_to.name,))
    raise
```

File: pyomo/contrib/mindtpy/util.py
Comment (jsiirola) [inline pyomo/contrib/mindtpy/util.py]: What about other domains like `NonPositiveReals`?  Maybe something more general:
```python
elif abs(var_val) <= config.zero_tolerance and 0 in v_to.domain:
    v_to.set_value(0)
```

File: pyomo/contrib/mindtpy/util.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/util.py]: OK

**Review by mrmundt — APPROVED** [review]

**Comment by ZedongPeng** [comment]
Hi @jsiirola . Do I need to further improve the code coverage of this PR? I have tried, but it's not very easy. 

> Hi John @jsiirola. Function `copy_var_list_values_from_solution_pool` will extract the variable values from the solution pool interface. However, I didn't find a way to set or change these values. For example, when using `gurobi_persistent`, the following error will show up.
> 
> > `AttributeError: Attribute 'Xn' cannot be set.`
> 
> They seem to be read-only. Therefore, I removed the redundant code in `copy_var_list_values_from_solution_pool` function, which will also improve the code coverage.

**Review by jsiirola — APPROVED** [review]

**Comment by jsiirola** [comment]
@ZedongPeng: Nope.  It has the two required reviews - I just missed that it got the second review.  @mrmundt could have merged it, but I am betting she was waiting for me to look at your last changes.

**Comment by ZedongPeng** [comment]
Thanks to both of you.  @jsiirola @mrmundt

**Pull request merged**
